pkg/ansible/proxy/cache_response.go

```diff
@@ -225,12 +226,22 @@ func (c *cacheResponseHandler) getListFromCache(r *requestfactory.RequestInfo, r
 	k.Kind = k.Kind + "List"
 	un := unstructured.UnstructuredList{}
 	un.SetGroupVersionKind(k)
-	err := c.informerCache.List(context.Background(), &un, clientListOpts...)
-	if err != nil {
-		// break here in case resource doesn't exist in cache but exists on APIserver
-		// This is very unlikely but provides user with expected 404
-		log.Info(fmt.Sprintf("cache miss: %v err-%v", k, err))
-		return nil, err
+	errChan := make(chan error, 1)
+	go func() {
+		err := c.informerCache.List(context.Background(), &un, clientListOpts...)
+		errChan <- err
+	}()
+
+	select {
+	case watchErr := <-errChan:
+		if watchErr != nil {
+			// break here in case resource doesn't exist in cache but exists on APIserver
+			// This is very unlikely but provides user with expected 404
+			log.Info(fmt.Sprintf("cache miss: %v err-%v", k, watchErr))
+			return nil, watchErr
+		}
+	case <-time.After(cacheEscacheEstablishmentTimeout):
+		return nil, fmt.Errorf("timeout establishing watch, commonly permissions of the controller are not sufficent")
```

Comment:
> Rather than using a channel, a goroutine, and a timer, a more idiomatic approach would be to use context.WithTimeout(context.Background(), cacheEstablishmentTimeout)
> 
> And then check the error from the Get() call to see if it is a context.DeadlineExceeded error.
> 
> Ditto for the informerCache.List() call above.
>

Reply:
required change

Reply:
Solved